File: 0143-reorder-list/0143-reorder-list.cpp
class Solution
{
public:
    ListNode *middleNode(ListNode *head)
    {
        ListNode *slow = head;
        ListNode *fast = head->next;
        while (fast && fast->next)
        {
            fast = fast->next->next;
            slow = slow->next;
        }

        return slow;
    }
    ListNode *reverseList(ListNode *secondHalf)
    {
        ListNode *current = secondHalf;
        ListNode *past = nullptr;
        while (current)
        {
            ListNode *next = current->next;
            current->next = past;
            past = current;
            current = next;
        }
        return past;
    }
    void reorderList(ListNode *head)
    {
        ListNode *middle = middleNode(head);
        ListNode *secondHead = reverseList(middle);
        ListNode *firstHead = head;
        while (secondHead && firstHead)
        {
            ListNode *next1 = firstHead->next;
            ListNode *next2 = secondHead->next;
            firstHead->next = secondHead;
            secondHead->next = next1;
            firstHead = next1;
            secondHead = next2;
        }
    }
};

File: 0143-reorder-list/0143-reorder-list.java
class Solution {
    public void reorderList(ListNode head) {

        ListNode secondHalfStart = splitListIntoToHalfs(head);
        ListNode head2 = reverseLinkedList(secondHalfStart);
        merge(head, head2);


    }


    private ListNode splitListIntoToHalfs(ListNode head) {
        ListNode fast = head;
        ListNode slow = head;
        while (fast != null && fast.next != null) {
            fast = fast.next.next;
            slow = slow.next;
        }
        ListNode
                secondHalfHead = slow.next;
        slow.next = null;
        return secondHalfHead;
    }

    private ListNode reverseLinkedList(ListNode head) {
        ListNode curr = head;
        ListNode prev = null;

        while (curr != null) {
            ListNode temp = curr.next;
            curr.next = prev;
            prev = curr;
            curr = temp;
        }

        return prev;
    }

    private ListNode merge(ListNode head1, ListNode head2) {
        ListNode curr1 = head1;
        ListNode curr2 = head2;
        while (curr1 != null && curr2 != null) {
            ListNode next1 = curr1.next;
            ListNode next2 = curr2.next;

            curr1.next = curr2;
            curr2.next = next1;

            curr1 = next1;
            curr2 = next2;
        }

        return head1;
    }


}